### ISSUE TITLE
exit and return default value is 0

### ISSUE THREAD

**Issue by lhoursquentin** [oilshell/oil#423]
Seems like `exit` without args is currently the same as `exit 0` when it should instead be the same as `exit "$?"` (same goes for `return`)

```bash
false
exit
```
and
```bash
f() {
  false
  return
}
f
```
Both "return" 0 instead of 1

**Comment by lhoursquentin** [comment]
Related to this one:

**Comment by lhoursquentin** [comment]
Related to this one:
```bash
$(false)
```
Also returns 0

**Comment by andychu** [comment]
Thanks, great report -- somehow I didn't know this :)  Fixed.

The `$(false)` issue is slightly different and related to `strict-argv`:

https://www.oilshell.org/release/0.6.0/doc/osh-manual.html#toc_4

Basically empty argv arrays return 0 in OSH, they don't take on the last command sub status.

Maybe it needs some more documentation, e.g. in the known differences doc.

**Comment by lhoursquentin** [comment]
Thanks for the fix!

Ha ok understood, my use case for `$(false)` was something like:
```bash
$(some_cmd | grep test) && some_other_cmd
```
Which now needs to be:
```bash
to_run=$(some_cmd | grep test) && $to_run && some_other_cmd
```
Agreed not a big deal.

I do see your point with empty argv arrays, but from a high level perspective I find it a bit weird that `a=$(false)` and `$(false)` have different return values.
Documenting this as a known difference would be great indeed!

**Comment by andychu** [comment]
@lhoursquentin 

Does this do the same thing?

```
ls | grep test >/dev/null && echo found
```

Or to be explicit about precedence:

```
{ ls | grep test > /dev/null; } && echo found
```

I wasn't aware of any "real" use cases for the exit code of `$(command-with-no-output)`.  I just saw it on the bash mailing list once as a curiosity.  If there is one I would reconsider it.

**Comment by andychu** [comment]
Yeah I don't understand your use case.  You want to run the output of `grep` as a command ?  Can you give an example in context?

**Comment by lhoursquentin** [comment]
Yes the idea was to run the output of `grep`.

My "real" use case is just a dirty hack that could be split into multiple pieces, that's why I was trying to give a simpler example but this made the whole thing even more confusing.

So I don't exactly **need** this specific form, I was just using it because it worked.

Anyways, here's a full working snippet which is very close to what I'm really doing (this is in `bash`):
```bash
$ cat mapping
{
  x: 1
  y: 2
}

$ sed -n 's/ *\(.*\): \(.*\)/declare \1=\2/p' mapping | grep '' # grep to return 1 for sed not returning anything
declare x=1
declare y=2

$ echo "$?"
0

$ sed -n 's/ *\(.*\): \(.*\)/declare \1=\2/p' /dev/null | grep ''

$ echo "$?"
1
```
And so:
```bash
$ $(sed -n 's/ *\(.*\): \(.*\)/declare \1=\2/p' mapping | grep '') && echo variables declared
variables declared

$ echo "$x"
1

$ $(sed -n 's/ *\(.*\): \(.*\)/declare \1=\2/p' /dev/null | grep '') && echo variables declared

$
```
(added additional empty lines for clarity)

As I said previously, this can be done in many different and cleaner ways, so this probably will not help to convince you, probably the opposite.

My only concern was that `a=$(false)` and `$(false)` returning different values seems inconsistent, I know these do different things, but instinctively I would expect them to return the same value.

**Comment by lhoursquentin** [comment]
This is getting further and further from the original point but on that same topic I would expect `declare a=$(false)` to return 1 too, especially since you treat assignment builtins as close as possible as the standard assignment, though in that case it would differ from all other shells, but this could make sense IMO.

I remember a thread on bug-bash ([link](https://lists.gnu.org/archive/html/bug-bash/2018-11/msg00077.html)) about builtin assignment error having special parsing rules but not being fatal in case of error, I guess this could be something to think about.

**Comment by andychu** [comment]
Yeah I think this area need some revisiting.  I did a `blame` on the code and I found this commit that added a hack to run Alpine's `abuild`.

https://github.com/oilshell/oil/commit/ba772f081056beb08fd25d8db4190076f25f20b2

I think this hack goes halfway only... It should probably be entirely compatible or be "rational" with an option for compatibility (`set -o posix` or something).

Let's make a new bug for it.
